Improve extraction of GitHub issue data

bockthom opened the pull request:
In this pull request, I add lots of improvements and take of more consistency of the issue data.

This pull request is related to se-passau/GitHubWrapper#25.
This pull request should only be merged as far as se-passau/GitHubWrapper#25 is approved, at least (or merged in the end).

Here are the main enhancements:

- Add review data, review comments, review requests, review dismiss events [94bab3a, 4eb6e20, 
b4283f9, 3862355, 16debcf]
- Unify GitHub users, that is, bring usernames and name/e-mail of commits together [9370972, 
973e2b9, 1faf340, 47e3ec8, 6adb906]
- Fix and enhance information for "assigned", "unassigned", "mentioned", "subscribed" events [
622b7f8, 8d66e45, b135036]
- Distinguish related commits in "add_link", "comitt_added", and "referenced_by" [39ff302, 2254fb9, 2b5e72b, 5602a7f ]
- Add a reversed event for referencing issues, that is, "referencedy_by" as opposite of "add_link" [c34842a, 2b5e72b]
- Remove duplicated events [b9a7182]
- Automatically add a "commented" event for each "created" event [18a65da, 066d875]

As soon as my enhancements are approved, I will upload an updated version of the diagram which shows all the events and their possible fields.

Please let me know if you have any suggestions on how to further improve the issue data.

bockthom committed:
Check content type in mbox-parsing only for lower-case characters

In the mbox-parsing, when checking whether the content-type of a message
contains "text", the content type needs to be set to lower case
characters before performing the check for "text" being part. Reason:
Usually the content type is, e.g., "text/plain", but there are lots of
e-mails whose content type is "TEXT/PLAIN".

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Add reviews and review comments from GitHub to issue data

Add reviews as events, initial review comments as comments, and review
comments as comments.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Unify GitHub users

When reformating the events of GitHub issues, update and unify users:
For instance, we can get actual names and e-mail addresses from
referenced commits for users for which we only have a username. To add
this information also for other events where those data is missing, a
dictionary of users is built and if some data is missing, the data from
the dictionary is used.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Fix pull-request commits and rename "has_commits" to "commit_added"

If a related-commit event does not contain a username, the commit was
just added to the pull request.

In addition, rename event "has_commit" to "commit_added" to more
concrete.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Fix initial comment of GitHub reviews

Due to a variable name conflict, two different comments were reduced to
one comment. This is fixed by renaming one of the conflicting variables.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Fix mention and subscribe events

Currently, we assume that mention and subscribe events take place at
exactly the same time as the comment in which they get mentioned.
However, there are examples (e.g., issue number 8834 of OpenSSL) in
which the mention and subscribe events take place one seconds later than
the actual comment they get mentioned in. Hence, if we do not find a
comment at exactly the same time, we have to look for a comment one
second earlier to be able to correctly fix the mentioned and subscribe
events' authors.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Extract requested reviewer for GitHub pull-request review requests

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Create dictionary of users from all issues

It is not enough to create the users dictionary just from the events of
one issue, as a user can only be active in another issue and will not
have it in the dictionary of an issue where the user is referenced.
Hence, the dictionary is created upon all issues.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Correctly distinguish referenced commits and commits added to a pull request

As the GitHubWrapper was adjusted to differentiate between referenced
commits and commits added to a pull request using a new means, this also
is adjusted when processing an issue.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Remove duplicated events

When someone merges a pull request, the merge commit originating from
the merge of the pull request causes two "add_link" commits, resulting
in two "add_link" events of the same user at exactly the same time
referencing the exactly the same commit in exactly the same pull
request. As there are two events for referencing the same commit, we
need to remove duplicates (but preserve the order of the events).

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Fix review information of GitHub pull requests

When a review on GitHub gets dismissed in a later event, the original
state of the review gets replaced by "dismissed". Hence, the state of
the review is wrong as it was not "dismissed" at the moment the review
was created. However, the "review_dismissed" event stores the original
state of the dismissed review. As a consequence, we have to replace the
state of the review by its original state. (To not omit the information
that the review was dismissed later on, the "event_info_2" field is set
to "later_dismissed".)

In addition, in the case that a review is manually dismissed, the person
who dismisses the review can provide a short dismissal message. If such
a message is available, we add an additional "commented" event for that
to the list of comments.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Use user information from authors of commits on GitHub

As the json file resulting from the GitHubWrapper does contain author
and committer information for all related commits, we can use this
information also for building our user dictionary, which is necessary to
merge GitHub usernames with real names and e-mail addresses resulting
from the git repository. As the actor of a related-commit event is the
committer in general, we only need to additionally add the author data.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Add additional "commented" event for the creation event of a GitHub issue

Adding an additional "commented" event for the creation of an issue is
necessary as a the creation comment can already mention other users. As
only "commented" events are used to fix authors and targets of
"mentioned" and "subscribe" events, we need to add the creation comment
also to the list of comments and, therefore, we add an additional
"commented" event.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Add dismissal comments before formating events

As dismissal comments can also mention other users, we need to add the
dismissal comments to the list of comments before formating events as
all the comment information is used to fix authors and targets of
"mentioned" and "subscribed" events.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Correctly extract "assigned" events of GitHub issues

In the GitHub issues data, the "user" of an "assigned" event actualy is
the assignee, which is passive in this event. Hence, set the assignee to
the ref_target column and extract the actual, active user from the
"assigner" field.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Handle "unassigned" and "review_request_removed" events

"unassigned" events behave as "assigned" events;
"review_request_removed" events behave as "review_requested" events.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Add commented event for creation event regardless whether it is not empty

Even if the creation event of an issue does contain an empty content,
add a commented event for this creation event.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Add "referenced_by" event to each issue which gets referenced by another issue

The list of related issues contains events in which a user has
referenced another issue within the current issue. In this commit, the
reversed event is also added: The referenced issue gets an event which
states that is was referenced in another issue.

As we cannot modify another issue while iterating over the list of
issues, the reversed events need to be temporarily stored in a temporary
list and added to the actual issue afterwards.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Fix dictionary of users by adding missing users

Add ref_target users also to dictionary of users as the could be missing
otherwise, which would result in key errors.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Distinguish whether a commit references an issue or the other way round

The event "add_link" is used when someone references another issue or
another commit in the current issue (e.g., in a comment).

The event "referenced_by" is used when someone references the current
issue within a commit message or in another issue.

For related issues, this was already fixed previously. Now, this is also
fixed for related commits, to stay consistent.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

clhunsen commented on issue_processing/issue_processing.py:
```diff
@@ -352,9 +352,34 @@ def reformat_events(issue_data):
         # re-format information of every event in the eventsList of an issue
         for event in issue["eventsList"]:
 
+            if (event["user"]["name"] == "" or event["user"]["name"] is None or
```
Can we refactor this into a separate lookup function? Then, we could pass `event["user"]` and `event["ref_target"]` and return the altered or unaltered user "object".

bockthom replied:
Done.

clhunsen commented on issue_processing/issue_processing.py:
```diff
@@ -120,6 +120,19 @@ def format_time(time):
+    """
+
+    new_time = datetime.strptime(time, "%Y-%m-%d %H:%M:%S") - timedelta(seconds = seconds)
+    return new_time.strftime("%Y-%m-%d %H:%M:%S")
```
Can we put this String constant into a file-local constant?

bockthom replied:
Yes, sounds like a good idea!

clhunsen reviewed (approved):
Wow, what an extension of the GitHub-related functionality. :open_mouth: I only have suggestions for refactorings, otherwise the code looks good to me.

bockthom committed:
Fix dictionary of users

Due to a copy-and-paste mistake, users in the dictionary which are
extracted from the 'ref_target' column are updated with values from the
'user' column, which is wrong, indeed. This is fixed now.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Apply minor improvements from review in PR #26

- Create a constant for the datetime string format.
- Extract updating of the user dictionary and lookups in the user
  dictionary to separate functions to avoid duplicated code.

Thanks to @clhunsen for these remarks.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Remove event "referenced" when target is "commit"

Up to now, we kept "referenced" events whose target is "commit".
Nevertheless, those events are duplicated as they also will appear in
the commits list. As we use the commits list to determine the correct
event type ("add_link", "referenced_by", "commit_added"), we don't need
the "referenced" event here any more. Therefore, such "referenced"
events will be removed, in the end.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom commented:
If have added some further improvements on this PR now.

Actually, I extracted the data from this PR for evaluating some special cases, resulting in some additional corner cases :wink: 

The initial comment of this PR leads to another problem which we have to think about how to deal with in the GitHubWrapper: In the initial comment, I have referenced issue 25 from the GitHubWrapper. However, by the GitHubWrapper, this is recognized as a link from codeface-extraction 26 (this PR) to codeface-extraction 25 (which actually is not referenced). Sorry for not referencing those things in this comment - that would lead to lots of references in the data of this issue, which I wanted to avoid for debugging reasons. To mitigate this issue, I suggest the following solutions to be solved in the GitHubWrapper: 
- Check if there is a space before `#` or this sign is at the beginning of a line. If not, check if the preceding prefix ends with the name of the repository (as far as this is available in the GitHubWrapper while issue extraction...)

bockthom committed:
Don't lookup user in the user dictionary if they have no username

If some users do not have a username (e.g., users originating from
commits which do not have a GitHub account), don't perform a lookup in
the user dictionary as there are only users with usernames in the
dictionary.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom commented:
I have run the issue extraction on the complete OpenSSL data and the result look good (at least, from what I have spot-checked). So, from my point of view, this PR is ready to merge if there are no further change requests. 
[Please postpone merging this PR until se-passau/GItHubWrapper#25 is merged.]

clhunsen commented:
Sound really good, @bockthom! :+1: The code looks good to me; @ecklbarb, @fehnkera, are there any remarks from your side?

bockthom commented:
I have updated the corresponding diagram now:
[issue_data_processing.pdf](https://github.com/se-passau/codeface-extraction/files/3450757/issue_data_processing.pdf)
Please have a look at it.

[Disclaimer: I only have mentioned the event types which occur in the recently extracted OpenSSL data, which is quite a lot (and even more than I have expected). There may be also other event types (which only rarely occur, though) in other projects.]

bockthom committed:
Fix missing user information in GitHub issue data

It can be the case that the user of a referenced commit event is "None".
In this case, we have to initialize "user" before updating it with
information from the user data of the commit.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Update copyright header

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>

bockthom committed:
Fix removing types and resolutions of GitHub issues

In some strange cases, in a GitHub issue, somebody has removed a label
that was never added before. If the removed label also is a "type" or an
"resolution", removing this type or resolution from the corresponding
list does not work as it is not in the list. So, only remove it if it is
in the list.

Signed-off-by: Thomas Bock <bockthom@fim.uni-passau.de>
